src/librustc/hir/map/mod.rs

```diff
@@ -506,6 +507,35 @@ impl<'hir> Map<'hir> {
         &self.forest.krate.attrs
     }
 
+    pub fn visit_module_item_likes<V>(&self, module: DefId, visitor: &mut V)
+        where V: ItemLikeVisitor<'hir>
+    {
+        let node_id = self.as_local_node_id(module).unwrap();
+
+        // Read the module so we'll be re-executed if new items
+        // appear immediately under in the module. If some new item appears
+        // in some nested item in the module, we'll be re-executed due to the reads
+        // in the loops below
+        self.read(node_id);
+
+        let module = &self.forest.krate.modules[&node_id];
+
+        for id in &module.items {
+            self.read(*id);
+            visitor.visit_item(&self.forest.krate.items[id]);
+        }
+
+        for id in &module.trait_items {
+            self.read(id.node_id);
+            visitor.visit_trait_item(&self.forest.krate.trait_items[id]);
+        }
+
+        for id in &module.impl_items {
+            self.read(id.node_id);
+            visitor.visit_impl_item(&self.forest.krate.impl_items[id]);
```

Comment:
I think it would be preferable to use `self.expect_impl_item(...)` here and in the other two cases. It might be slightly less efficient (but probably not measurably so) but it would remove the danger of the dep-tracking logic here drifting apart from other places.

But I'll let you decide. It's not a big deal.